### ISSUE TITLE
TLS based two way authentication

### ISSUE THREAD

**Issue by jliu1970** [kerryjiang/WebSocket4Net#94]
Hi Kerry,

Is there any way to implement a client that supports client certificate for a TLS based two way authentication?

Thanks
Jimmy

**Comment by kerryjiang** [comment]
Please reference the latest beta version and set certificate with the Security property of the websocket instance.

**Comment by jliu1970** [comment]
Thanks for the quick response, Kerry. I will study it.
Jimmy

**Comment by jliu1970** [comment]
Hi Kerry,

I looked through the 0.15.0-Beta7 version, but could not find a way to set the client certificate that it will use to authenticate itself with server. Any more specific hint? 

I appreciated your help
Jimmy

**Comment by jliu1970** [comment]
I saw that CreateSecureTCPSession sets the SecurityOption for the client. SecurityOption seems to be the place to host the X509Certificates. But I could not figure out access the securityoption object from my application. Please suggest.

**Comment by jliu1970** [comment]
Or would that be possible to add setter to WebSocket.Security property, if I understand the design correctly.

Thanks
Jimmy

**Comment by kerryjiang** [comment]
I think you can access the Security property and add client certificate into the Certificates like the code below without creating a new SecurityOption instance:

websocket.Security.Certificates.Add(xxx);

Let me know if it works for you.

**Comment by jliu1970** [comment]
Thanks for the advice, Kerry.
Jimmy

**Comment by jliu1970** [comment]
Kerry, 
Your advice works. Is there anywhere I can get a strong named assembly? 
Thanks again,
Jimmy

**Comment by kerryjiang** [comment]
Do you mean the WebSocket4Net's assembly? I thought it was signed. Do you use .NET core version?

**Comment by jliu1970** [comment]
Kerry,

Thanks for the quick response.

 I used Nuget to download the pre-release version Beta8. Then in Developer Command Prompt for VS2017, I ran the command of "sn -v" against the ....\WebSocket4Net.0.15.0-beta8\lib\net45\WebSocket4Net.dll, which returned that "... WebSocket4Net.dll does not represent a strongly named assembly".

Sorry if my description may not be clear, I have been a Java developer and switched to C# at the end of last year.

Thanks for your help,
Jimmy

**Comment by kerryjiang** [comment]
If so, I have to check this issue today.

**Comment by jliu1970** [comment]
Thanks again, Kerry.
Jimmy

**Comment by kerryjiang** [comment]
Fixed the signing issue. Could you check out the latest version from the Nuget?

**Comment by jliu1970** [comment]
Thanks, Kerry. It works. You are the best.
Jimmy

**Comment by jliu1970** [comment]
Hi Kerry,

Sorry, my question keeps coming, and this one is probably on SuperSocket.ClientEngine.

The issue we are facing is that we have to support server certificate validation at client side using server certificate that is not stored in the trusted store (somehow, IA team deems that store server certificate in trusted store is not secure). I saw codes in SslStreamTcpSession.ValidateRemoteCertificate() method that in a !NETSTANDARD environment, it seems that we can implement this through a  System.Net.ServicePointManager.ServerCertificateValidationCallback delegate at the client side. But in the case of .net environment, is there anyway we can provide some customization?

Please advice.

Thanks
Jimmy

**Comment by kerryjiang** [comment]
You should be able to use ServicePointManager.ServerCertificateValidationCallback in your .NET 4.0/4.5 project. Because !NETSTANDARD means, the framework is a normal framework like .net 4.0 and .net 4.5 instead of netstandard of .net core.

**Comment by jliu1970** [comment]
Thanks again for the advice.

Have a nice weekend,
Jimmy

**Comment by kerryjiang** [comment]
Does it work for you?

**Comment by jliu1970** [comment]
Sorry, Kerry. Forgot to update you. 
We are waiting for the server side code update to test it out. By any chance, are you aware of any public available server that we can test our client?
I will be on a vacation to China later this week, will update you once I come back and test it out.
Thanks
Jimmy

**Comment by jliu1970** [comment]
Hi Kerry,

Sorry for the delayed update.  I just came back from vacation today.

We had a few churns on the requirements, and ended up implementing a X509 certificate based mutual authentication + digest access authentication. The digest was passed in through customHeaderItems.

We ran the testing just now with the server. and observed that the websocket was opened successfully and the communication between our client and server worked as expected.

Thanks for all the advices on this issue.
Jimmy
